docs/architecture/adr-010-enable-standalone-ics02-integration.md

```diff
@@ -0,0 +1,372 @@
+developers who wish to integrate only the implemented client states under the
+ibc-rs to their IBC infra. In other words, this enhancement does not mandate
+hosts to implement all the IBC layers, such as connection, channel, or any of
+the IBC applications, when only light client support is required.
+
+As a result, this ADR sets the stage for several potential use cases, including:
+
+1. **Streamlines CosmWasm light client implementation**
+    - Since the refined structure allows for the independent integration of
+      ICS-02 client traits into the storage, it becomes straightforward to
+      implement light clients living in ibc-rs as CosmWasm contracts, which
+      empowers simpler Wasm client integration with ibc-go driven chains.
+
+2. **Enables client integration with various IBC implementations**
+    - Any IBC implementation, whether it is based on ibc-rs or a fork of ibc-rs,
+      will have this capability to incorporate the latest light client
+      implementations by ibc-rs. So that even if an IBC implementation diverges
+      from the ibc-rs framework, hosts can benefit from the latest version of
+      light client implementation. This is important from the security
+      standpoint, broadens the applicability of ibc-rs, and facilitates
+      collaboration with other IBC-adjacent solutions.
+
+As part of this ADR and to minimize breaking changes later, we aim to address
+the following client-relevant API deficiencies as well:
+
+1. **Refining access to validation vs execution methods**
+    - So far, unauthorized access to client execution methods [has been
+      possible](https://github.com/cosmos/ibc-rs/blob/4421b0dcd6eab6c914b8c5a7d45817b1edb6b2ce/ibc-core/ics24-host/src/context.rs#L33)
+      under the `ValidationContext`. This will be rectified to ensure that only
+      the `ClientState` validation method is callable within the validation
+      contexts. This enhances the precision of access control, aligning it more
+      closely with the intended functionality.
+
+2. **Bringing more clarity to how context APIs are related**
+    - The ambiguous relationship between the main context APIs at ICS-24, ICS-02
+      client APIs, and client-specific APIs can get cleared. This ADR aims for a
+      more transparent relationship between these contexts, making it easier to
+      understand their interactions and roles. For instance some client relevant
+      methods are located under the client contexts, while others are under the
+      main (ICS-24) contexts.
+
+3. **Minimizing associated types’ declarations**
+    - Our current APIs pose a challenge by requiring users to introduce multiple
+      yet the same associated types like AnyClientState, AnyConsensusState, etc
+      across different layers. A more efficient solution involves introducing
+      such types in fewer places, providing better access to the traits across
+      different layers.
```

Comment:
Can you please comment here on what features we need from rustc to improve this later?

- associated_type_bounds
- associated_type_defaults
- implied_bounds

Reply:
[034604a](https://github.com/cosmos/ibc-rs/pull/1116/commits/034604aca6663a554ea1704b01fbbd6dd899789f)